File: packages/distance/src/token/naive.ts
/**
 * Naive string similarity measures: prefix, suffix, length.
 *
 * Time: O(min(m, n)) for prefix/suffix, O(1) for length
 */

/**
 * Compute prefix similarity between two strings.
 *
 * prefix(a, b) = commonPrefixLength / max(|a|, |b|)
 *
 * @param a - First string
 * @param b - Second string
 * @returns Prefix similarity in [0, 1]
 */
export function prefix(a: string, b: string): number {
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 1;

  let commonLen = 0;
  while (
    commonLen < a.length &&
    commonLen < b.length &&
    a.charCodeAt(commonLen) === b.charCodeAt(commonLen)
  ) {
    commonLen++;
  }

  return commonLen / maxLen;
}

/**
 * Compute suffix similarity between two strings.
 *
 * suffix(a, b) = commonSuffixLength / max(|a|, |b|)
 *
 * @param a - First string
 * @param b - Second string
 * @returns Suffix similarity in [0, 1]
 */
export function suffix(a: string, b: string): number {
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 1;

  let commonLen = 0;
  const aEnd = a.length;
  const bEnd = b.length;
  while (
    commonLen < aEnd &&
    commonLen < bEnd &&
    a.charCodeAt(aEnd - 1 - commonLen) === b.charCodeAt(bEnd - 1 - commonLen)
  ) {
    commonLen++;
  }

  return commonLen / maxLen;
}

/**
 * Compute length-based similarity between two strings.
 *
 * length(a, b) = 1 - |len(a) - len(b)| / max(len(a), len(b))
 *
 * @param a - First string
 * @param b - Second string
 * @returns Normalized length similarity in [0, 1]
 */
export function length(a: string, b: string): number {
  const maxLen = Math.max(a.length, b.length);
  if (maxLen === 0) return 1;
  return 1 - Math.abs(a.length - b.length) / maxLen;
}

File: packages/distance-wasm/src/token/naive.rs
/// Naive string similarity: prefix, suffix, length.
///
/// Time: O(min(m, n)) for prefix/suffix, O(1) for length

pub fn prefix(a: &str, b: &str) -> f64 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let max_len = a_bytes.len().max(b_bytes.len()) as f64;
    if max_len == 0.0 { return 1.0; }

    let min_len = a_bytes.len().min(b_bytes.len());
    let mut common = 0usize;
    while common < min_len && a_bytes[common] == b_bytes[common] {
        common += 1;
    }
    common as f64 / max_len
}

pub fn suffix(a: &str, b: &str) -> f64 {
    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let max_len = a_bytes.len().max(b_bytes.len()) as f64;
    if max_len == 0.0 { return 1.0; }

    let min_len = a_bytes.len().min(b_bytes.len());
    let mut common = 0usize;
    while common < min_len
        && a_bytes[a_bytes.len() - 1 - common] == b_bytes[b_bytes.len() - 1 - common]
    {
        common += 1;
    }
    common as f64 / max_len
}

pub fn length(a: &str, b: &str) -> f64 {
    let max_len = a.len().max(b.len()) as f64;
    if max_len == 0.0 { return 1.0; }
    1.0 - (a.len() as f64 - b.len() as f64).abs() / max_len
}
